Task: Spring Boot Portfolio Project — Part 6. Executor: Oleksii Bashkirov.

OleksiiBashkirov opened the pull request:
feat: Incorporate Spring Security with Role-based Access Control

- Added `spring-boot-starter-security` to the project.
- Created `SecurityConfig` class with `PasswordEncoder` and `SecurityFilterChain` beans.
- Implemented `UserDetails` in `User` class and `UserDetailsService`.
- Made `/auth/registration` publicly available.
- Added `Role` entity and `RoleRepository`.
- Added `Set<Role> roles` field to `User` entity.
- Updated Liquibase changesets for roles and user roles.
- Secured endpoints with `@PreAuthorize` based on roles.

OleksiiBashkirov committed:
Fixed security configuration and liquibase issue

OleksiiBashkirov committed:
Fixed checkstyle errors

MentorMateAcademy commented on src/main/java/mate/academy/repository/RoleRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
```suggestion
```

OleksiiBashkirov committed:
- Removed the redundant Repository annotation in the RoleRepository interface.

Elena-Bruyako commented on pom.xml:
```diff
@@ -101,6 +101,12 @@
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.3.0</version>
```
```suggestion
```

Elena-Bruyako commented on src/main/java/mate/academy/config/SecurityConfig.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
```
```suggestion
@Configuration
@EnableMethodSecurity
```

Elena-Bruyako commented on src/main/java/mate/academy/config/SecurityConfig.java:
```diff
@@ -0,0 +1,39 @@
+                .requestMatchers(HttpMethod.PUT, "/api/books/**").hasRole("ADMIN")
+                .requestMatchers(HttpMethod.DELETE, "/api/books/**").hasRole("ADMIN")
+                .anyRequest().authenticated()
+            )
```
add userDetailsService

Elena-Bruyako commented on src/main/java/mate/academy/model/User.java:
```diff
@@ -31,4 +40,47 @@ public class User {
 
     private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
```suggestion
    @ManyToMany
```

Elena-Bruyako commented on src/main/java/mate/academy/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String username)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found."));
```
```suggestion
{
        return userRepository.findByEmail(username)
                .orElseThrow(() -> new EntityNotFoundException("User not found."));
```

Elena-Bruyako commented on src/main/java/mate/academy/service/impl/UserServiceImpl.java:
```diff
@@ -24,6 +26,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
 
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
```
find role by name add it to roles

Elena-Bruyako reviewed (changes requested):
see comments

OleksiiBashkirov committed:
- Updated pom.xml file;
- Added annotations in SecurityConfig;
- Included userDetailsService in SecurityConfig;
- Removed EAGER fetch type for roles in User entity;
- Updated CustomUserDetailsService to use EntityNotFoundException;
- Implemented role assignment in UserServiceImpl during user registration.

sarakhmen commented on src/main/java/mate/academy/security/CustomUserDetailsService.java:
```diff
@@ -14,8 +19,22 @@ public class CustomUserDetailsService implements UserDetailsService {
+        User user = userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + username));
+
+        System.out.println("User found: " + user.getEmail());
+        System.out.println("User roles: " + user.getRoles());
```
remove these debug messages

OleksiiBashkirov replied:
Thank you for pointing that out. 
I had accidentally left those debug messages in, but I have now removed them.
Thank you for your guidance!

sarakhmen reviewed (changes requested):
let`s improve your sokution

OleksiiBashkirov committed:
- Removed debug messages.

Elena-Bruyako commented on src/main/java/mate/academy/config/SecurityConfig.java:
```diff
@@ -0,0 +1,60 @@
+
+@Configuration
+@EnableWebSecurity
+@EnableMethodSecurity
```
Do you need both of these annotations `@EnableWebSecurity and @EnableMethodSecurity`?

Elena-Bruyako commented on src/main/java/mate/academy/mapper/BookMapper.java:
```diff
@@ -6,7 +6,9 @@
 import org.mapstruct.MappingTarget;
+import org.springframework.stereotype.Component;
 
+@Component
```
```suggestion
```

Elena-Bruyako commented on src/main/java/mate/academy/model/User.java:
```diff
@@ -31,4 +39,47 @@ public class User {
 
     private String shippingAddress;
+
+    @ManyToMany()
```
```suggestion
    @ManyToMany
```

Elena-Bruyako commented on src/main/java/mate/academy/security/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String username)
+            throws EntityNotFoundException {
+
+        User user = userRepository.findByEmail(username)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + username));
+
+        List<SimpleGrantedAuthority> authorities = user.getRoles().stream()
+                .map(role -> new SimpleGrantedAuthority(role.getName().name()))
+                .collect(Collectors.toList());
+
+        return new org.springframework.security.core.userdetails.User(
+                user.getEmail(),
+                user.getPassword(),
+                authorities
+        );
```
```suggestion
           {

        return userRepository.findByEmail(username)
                .orElseThrow(() -> new EntityNotFoundException("User not found: " + username));
```

Elena-Bruyako commented on src/main/java/mate/academy/service/impl/UserServiceImpl.java:
```diff
@@ -24,6 +30,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        Role userRole = roleRepository.findByName(Role.RoleName.ROLE_USER)
+                        .orElseThrow(() -> new RegistrationException("Role not found."));
+        user.setRoles(Collections.singleton(userRole));
```
or you can use Set.of()

Elena-Bruyako reviewed (changes requested):
see comments

OleksiiBashkirov committed:
1. Updated SecurityConfig.java:
   - Removed the @EnableWebSecurity annotation.

2. Updated BookMapper.java:
   - Removed the @Component annotation.

3. Updated User.java:
   - Corrected @ManyToMany annotation formatting.

4. Updated CustomUserDetailsService.java:
   - Refactored the "loadUserByUsername" method.

5. Updated UserServiceImpl.java:
   - Replaced Collections.singleton with Set.of for setting user roles.

Elena-Bruyako reviewed (approved):
well done

OleksiiBashkirov commented:
Please disregard the last pull request in this theme made through the mate.academy platform. 
Due to an accidental click, I submitted changes meant for the "JWT" theme under the "Spring Boot Security" theme. 
Sorry about that!
